browser/ui/webui/brave_rewards/tip_panel_handler.cc

```diff
@@ -0,0 +1,177 @@
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/browser_tabstrip.h"
+#include "content/public/common/url_constants.h"
+#include "ui/gfx/geometry/size.h"
+#include "url/gurl.h"
+#include "url/url_constants.h"
+
+namespace brave_rewards {
+
+namespace {
+
+TipPanelCoordinator* GetCoordinator(Profile* profile) {
+  DCHECK(profile);
+  if (auto* browser = chrome::FindLastActiveWithProfile(profile)) {
+    return TipPanelCoordinator::FromBrowser(browser);
+  }
+  return nullptr;
+}
+
+std::string GetRequestedPublisherID(Profile* profile) {
+  if (auto* coordinator = GetCoordinator(profile)) {
+    return coordinator->publisher_id();
+  }
+  return "";
+}
+
+gfx::Size GetCurrentBrowserSize(Profile* profile) {
+  if (auto* coordinator = GetCoordinator(profile)) {
+    return coordinator->browser_size();
+  }
+  return gfx::Size();
+}
+
+}  // namespace
+
+TipPanelHandler::TipPanelHandler(
+    mojo::PendingRemote<mojom::TipPanel> banner,
+    mojo::PendingReceiver<mojom::TipPanelHandler> receiver,
+    base::WeakPtr<ui::MojoBubbleWebUIController::Embedder> embedder,
+    Profile* profile)
+    : receiver_(this, std::move(receiver)),
+      banner_(std::move(banner)),
+      embedder_(embedder),
+      profile_(profile),
+      rewards_service_(RewardsServiceFactory::GetForProfile(profile)),
+      publisher_id_(GetRequestedPublisherID(profile)) {
+  DCHECK(rewards_service_);
+}
+
+TipPanelHandler::~TipPanelHandler() = default;
+
+void TipPanelHandler::ShowUI() {
+  if (embedder_) {
+    embedder_->ShowUI();
+  }
+}
+
+void TipPanelHandler::CloseUI() {
+  if (embedder_) {
+    embedder_->CloseUI();
+  }
+}
+
+void TipPanelHandler::GetBrowserSize(GetBrowserSizeCallback callback) {
+  gfx::Size size = GetCurrentBrowserSize(profile_);
+  std::move(callback).Run(size.width(), size.height());
+}
+
+void TipPanelHandler::GetRewardsParameters(
+    GetRewardsParametersCallback callback) {
+  DCHECK(rewards_service_);
+  rewards_service_->GetRewardsParameters(std::move(callback));
+}
+
+void TipPanelHandler::GetBalance(GetBalanceCallback callback) {
+  DCHECK(rewards_service_);
+
+  auto fn = [](GetBalanceCallback callback, FetchBalanceResult result) {
+    if (!result.has_value() || !result.value()) {
+      std::move(callback).Run(nullptr);
+    } else {
+      std::move(callback).Run(std::move(result.value()));
+    }
+  };
+
+  rewards_service_->FetchBalance(base::BindOnce(fn, std::move(callback)));
+}
+
+void TipPanelHandler::GetBanner(GetBannerCallback callback) {
+  DCHECK(rewards_service_);
+
+  if (publisher_id_.empty()) {
+    std::move(callback).Run(nullptr);
+    return;
+  }
+
+  rewards_service_->GetPublisherBanner(publisher_id_, std::move(callback));
+}
+
+void TipPanelHandler::GetMonthlyContributionSet(
+    GetMonthlyContributionSetCallback callback) {
+  DCHECK(rewards_service_);
+
+  auto fn = [](GetMonthlyContributionSetCallback callback,
```

Comment:
Can we convert these lambdas into `*Callback` members?

Reply:
They don't require any weak self, so I think they are easier to read as lambdas?

Reply:
I think these flows read better if they go top-to-bottom (which is how we have them in most places). `private static` functions would be a decent alternative. They help organize our code better than inline callback definitions in my opinion.

Reply:
I'd like to leave them as-is.